src/main/java/net/minecraftforge/client/gui/config/widgets/ConfigGuiWidget.java

```diff
@@ -0,0 +1,716 @@
+                }
+            });
+            this.editBox.active = !specificationData.isSynced();
+            this.editBox.setEditable(this.editBox.active);
+        }
+
+        @Override
+        public boolean isValid()
+        {
+            return isValid(this.editBox.getValue());
+        }
+
+        @Override
+        public Object currentValue()
+        {
+            try
+            {
+                return parser.apply(this.editBox.getValue());
+            }
+            catch (NumberFormatException e)
+            {
+                return this.editBox.getValue();
+            }
+        }
+
+        @Override
+        public void resetToDefault()
+        {
+            this.editBox.setValue(this.valueSpec.getDefault().toString());
+            this.valueManager.setter().accept(currentValue());
+        }
+
+        @Override
+        public void resetToInitial()
+        {
+            this.editBox.setValue(this.valueManager.initial().get().toString());
+            this.valueManager.setter().accept(currentValue());
+        }
+
+        @Override
+        public Component getError()
+        {
+            return this.valueSpec.getError(currentValue());
+        }
+
+        @Override
+        public List<? extends NarratableEntry> narratables()
+        {
+            return Collections.singletonList(this.editBox);
+        }
+
+        @Override
+        public void render(final PoseStack poseStack, final int top, final int left, final int maxWidth,
+                           final int maxHeight, final int mouseX, final int mouseY, final boolean isHovered,
+                           final float partialTick)
+        {
+            this.editBox.setWidth(maxWidth - 4);
+            this.editBox.setHeight(maxHeight);
+            this.editBox.x = left + 2;
+            this.editBox.y = top;
+            this.editBox.render(poseStack, mouseX, mouseY, partialTick);
+        }
+
+        @Override
+        public @NotNull List<FormattedCharSequence> getTooltip()
+        {
+            return Collections.emptyList();
+        }
+
+        @Override
+        public @NotNull List<? extends GuiEventListener> children()
+        {
+            return Collections.singletonList(this.editBox);
+        }
+
+        @Override
+        public @NotNull NarrationPriority narrationPriority()
+        {
+            return NarrationPriority.HOVERED;
+        }
+
+        private boolean isValid(final String value)
+        {
+            try
+            {
+                return isValid(parser.apply(value));
+            }
+            catch (NumberFormatException e)
+            {
+                return false;
+            }
+        }
+
+        private boolean isValid(T value)
+        {
+            return this.valueSpec.test(value);
+        }
+    }
+
+    /**
+     * Default implementation for a {@link ConfigGuiWidget} for a {@link ForgeConfigSpec.IntValue}.
```

Comment:
As all these number widgets seem to have the same impl (their sub classes don't have any special overrides), should we maybe instead just expose the factories and have the factory create a new number widget instance with the corresponding parser rather than having to have a subclass AND factory for each one.